intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/staticanalysis/IntelliJMarkupModelListener.java

```diff
@@ -0,0 +1,220 @@
+import org.jetbrains.annotations.NotNull;
+import org.joda.time.DateTime;
+import org.joda.time.Seconds;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static nl.tudelft.watchdog.core.logic.ui.listeners.staticanalysis.StaticAnalysisMessageClassifier.IDE_BUNDLE;
+
+public class IntelliJMarkupModelListener extends CoreMarkupModelListener implements MarkupModelListener, Disposable {
+
+    static {
+        IDE_BUNDLE.createPatternsForKeysInBundle("messages.InspectionsBundle");
+        IDE_BUNDLE.createPatternsForKeysInBundle("com.siyeh.InspectionGadgetsBundle");
+
+        IDE_BUNDLE.sortList();
+    }
+
+    private final Document document;
+    private final TrackingEventManager trackingEventManager;
+    private final com.intellij.openapi.editor.Document intellijDocument;
+
+    private final Set<Warning<RangeHighlighterEx>> generatedWarnings;
+    private final Set<Warning<RangeHighlighterEx>> warnings;
+    private final Map<RangeHighlighterEx, DateTime> timeMapping;
+
+    private IntelliJMarkupModelListener(Document document, TrackingEventManager trackingEventManager, com.intellij.openapi.editor.Document intellijDocument) {
+        this.document = document;
+        this.trackingEventManager = trackingEventManager;
+        this.intellijDocument = intellijDocument;
+
+        generatedWarnings = new HashSet<>();
+        warnings = new HashSet<>();
+        timeMapping = new WeakHashMap<>();
+    }
+
+    /**
+     * Create a new {@link IntelliJMarkupModelListener} for an editor. This listener is only attached after the
+     * {@link DaemonCodeAnalyzer} has finished analyzing this file.
+     *
+     * The listener is attached to the {@link com.intellij.openapi.editor.markup.MarkupModel} of the document,
+     * which contains all {@link RangeHighlighterEx}s that represent the Static Analysis warnings.
+     *
+     * @param project The project the file exists in.
+     * @param disposable The disposable to clean up the listeners and any potential {@link MessageBusConnection}.
+     * @param editor The editor of the document.
+     * @param trackingEventManager The manager that can process all the events generated.
+     * @return A newly initiated listener that will later be attached to the {@link com.intellij.openapi.editor.markup.MarkupModel} of the document
+     */
+    public static IntelliJMarkupModelListener initializeAfterAnalysisFinished(
+            Project project, Disposable disposable, Editor editor, TrackingEventManager trackingEventManager) {
+
+        final com.intellij.openapi.editor.Document intellijDocument = editor.getDocument();
+        final IntelliJMarkupModelListener markupModelListener = new IntelliJMarkupModelListener(DocumentCreator.createDocument(editor), trackingEventManager, intellijDocument);
+
+        // We need to run this in smart mode, because the very first time you start your editor, it is very briefly
```

Comment:
Short explanation of smart vs. dumb mode? :-)

Reply:
Done